### PR TITLE
improvements to material code and material creation

### PR DESCRIPTION
**Pull request by pixelflinger**
- `ProgramBuilder` is a relic, rename to `Program`
- use the enum instead of ints everywhere for `ShaderModel` and `ShaderStage`
- Improve how we cache shared shaders:

Some shaders can be shared across all materials (e.g. the depth 
shaders). We use the filament default material as the "source" of
the cache, but until now we relied on an a priori knowledge of which
variants were present in the default material.

With this change, we now query once the list of variants (of interest)
in the default material and reuse that list for caching these variants
later.

This is better because the cached variants are now entirely driven by
the default material (which they depend on anyways). This is also faster
because we don't need to query which variant we need each time we create
a material.

### CODE REVIEW

**Review by romainguy — APPROVED** [review]

File: filament/src/details/Material.cpp [hunk 1]
```diff
@@ -358,6 +368,23 @@ BufferInterfaceBlock::FieldInfo const* FMaterial::reflect(
...
+    filaflat::MaterialChunk const& materialChunk{ mMaterialParser->getMaterialChunk() };
+
+    variants.reserve(materialChunk.getShaderCount());
+    materialChunk.visitShaders([&variants](
+            ShaderModel model, Variant variant, ShaderStage stage) {
+        variants.push_back(variant);
+    });
+    std::sort(variants.begin(), variants.end(),
+            [](Variant lhs, Variant rhs) { return lhs.key < rhs.key; });
+    std::unique(variants.begin(), variants.end());
```
Comment (bejado) [inline filament/src/details/Material.cpp:383]: Do we need to resize `variants` after calling `std::unique`?

File: filament/src/details/Material.cpp [hunk 2]
```diff
@@ -358,6 +368,23 @@ BufferInterfaceBlock::FieldInfo const* FMaterial::reflect(
     return mUniformInterfaceBlock.getFieldInfo(name);
 }
 
+utils::FixedCapacityVector<Variant> FMaterial::getVariants() const noexcept {
```
Comment (bejado) [inline filament/src/details/Material.cpp:371]: Are we using this anywhere?

File: filament/src/details/Material.cpp (hunk 1 above)
Comment (pixelflinger) [reply filament/src/details/Material.cpp:383]: uh. yes, I forgot to upload it. thanks!

Comment (pixelflinger) [reply filament/src/details/Material.cpp:383]: shoot I forgot to remove that code. thanks!

**Review by poweifeng — APPROVED** [review]

**Pull request merged**